man/man3/getopt.3
.\" Automatically generated by Pod::Man 4.10 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "STDIN 1"
.TH "getopt" 3 "2023-01-01" "md2man v0.1.0" "Declarative getopt (long) parser"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "GETOPT-BASH"
.IX Header "GETOPT-BASH"
Declarative Getopt Parser for Bash
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&    #!/bin/bash
\&
\&    source getopt.bash "\e
\&    getopt_default=(\-\-help)
\&
\&    my\-app <options...> <arguents...>
\&
\&    See \*(Aqman my\-app\*(Aq for more help.
\&
\&    Options:
\&    \-\-
\&    b,bar       Bar bar bar
\&    q,quux=     Quux quux
\&
\&    xxx         Xxx xxx
\&    yyy=        Yyy yyy
\&
\&    help        Show help
\&    version     Print my\-app version
\&    "
\&
\&    getopt "$@"
\&    set \-\- "${args[@]}"
\&
\&    if $option_bar || $option_xxx; then
\&      echo \*(AqOptions:\*(Aq
\&      echo "quux = $option_quux"
\&      echo "yyy = $option_yyy"
\&      echo \*(AqArguments:\*(Aq
\&      printf "\- %s\en" "$@"
\&    fi
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Git has a little known but very cool getopt parser built in. The Git command is:
.PP
.Vb 2
\&    git rev\-parse \-\-parseopt \-\- <command\-line\-args> \e
\&      <<<"<special\-format\-getopt\-spec\-text>"
.Ve
.PP
You pipe in a special format that looks like a help text describing the command. If you ask for \f(CW\*(C`\-\-help\*(C'\fR (or the parser fails to parse the options), it will print a slightly reformatted version of that spec as a help message.
.PP
This library makes that Git parser even more powerful and easy to use.
.SH "PARLANCE"
.IX Header "PARLANCE"
.IP "\(bu" 4
options (or opts)
.Sp
Options are the parameters like \f(CW\*(C`\-h\*(C'\fR, \f(CW\*(C`\-\-help\*(C'\fR, \f(CW\*(C`\-f <file\*(C'\fR>, \f(CW\*(C`\-\-file <file\*(C'\fR> or \f(CW\*(C`\-\-file=<file\*(C'\fR. Note that option names are case sensitive. ie >\-f\f(CW\*(C`and\*(C'\fR\-F` are diffent options.
.IP "\(bu" 4
arguments (or args)
.Sp
Arguments are the (zero or more) strings that (usually) follow the options.
.IP "\(bu" 4
long option
.Sp
An option of the form \f(CW\*(C`\-\-name\*(C'\fR or \f(CW\*(C`\-\-some\-thing\*(C'\fR. Two dashes and two or more letters in the name. The names should be lowercase, start with a letter and be composed of letters, numbers and dash separators. An \f(CW\*(C`=\*(C'\fR or a space may be used to separate the option from its value. ie \f(CW\*(C`\-\-file=abc.txt\*(C'\fR or \f(CW\*(C`\-\-file abc.txt\*(C'\fR.
.IP "\(bu" 4
short option
.Sp
An option of the form \f(CW\*(C`\-n\*(C'\fR or \f(CW\*(C`\-F\*(C'\fR. A single dash and a single character for the name. The character may be a lower or upper cased letter. If the option has a value, it is separated by a space: \f(CW\*(C`\-f abc.txt\*(C'\fR.
.IP "\(bu" 4
flag option
.Sp
Flag options never have a value parameter. The values returned by getopt-bash are \f(CW\*(C`false\*(C'\fR (default if the option is not present) and \f(CW\*(C`true\*(C'\fR if the option \fIis\fR present.
.IP "\(bu" 4
value option
.Sp
Value options always have a value parameter. This getopt-bash library returns the value specified or the empty string if it is not present in the parsed parameters.
.IP "\(bu" 4
dual option
.Sp
A dual option can be used alone like a flag option, or with a value.
.IP "\(bu" 4
getopt spec
.Sp
A multiline string that defines how getopt will behave. See \*(L"Spec Structure\*(R" below.
.IP "\(bu" 4
getopt config
.Sp
The getopt library has a number of config variables you can set to get custom behavior. These variables can be set as part of the getopt spec at the top. They can also be set using normal Bash variable assignments as well.
.SH "USAGE"
.IX Header "USAGE"
The \f(CW\*(C`getopt\-bash\*(C'\fR package provides a single library file \f(CW\*(C`lib/getopt.bash\*(C'\fR. If you \f(CW\*(C`source getopt.bash\*(C'\fR it provides a function called \f(CW\*(C`getopt\*(C'\fR.
.PP
The \f(CW\*(C`getopt\*(C'\fR function requires a getopt spec, and there are 3 ways to provide it:
.ie n .IP "1. As the ""getopt_spec"" config variable." 4
.el .IP "1. As the \f(CWgetopt_spec\fR config variable." 4
.IX Item "1. As the getopt_spec config variable."
.PD 0
.ie n .IP "1. As an argument to ""source getopt.bash ""$spec""""." 4
.el .IP "1. As an argument to \f(CWsource getopt.bash ``$spec''\fR." 4
.IX Item "1. As an argument to source getopt.bash ""$spec""."
.ie n .IP "1. As \s-1STDIN\s0 to the ""getopt"" function." 4
.el .IP "1. As \s-1STDIN\s0 to the \f(CWgetopt\fR function." 4
.IX Item "1. As STDIN to the getopt function."
.PD
.PP
The \f(CW\*(C`getopt\*(C'\fR function takes the arguments you want to parse (usually \f(CW$@\fR) as its arguments.
.PP
The function will (by default) set each option in a variable called \f(CW\*(C`option_<option\-name\*(C'\fR> and put any arguments into an array variable called \f(CW\*(C`args\*(C'\fR.
.PP
The declared option variables will default to \f(CW\*(C`false\*(C'\fR for flag options and \f(CW\*(Aq\*(Aq\fR for value options. Actually the default for a value option is an empty array, but when used in a scalar context it will return the first value or \f(CW\*(Aq\*(Aq\fR if the array is empty.
.PP
Any remaining words (not options) are placed in the array variable \f(CW\*(C`args\*(C'\fR.
.PP
If a value option is specified multiple times, the option variable will be an array with all the values in the order they were used.
.PP
The variable \f(CW\*(C`option_count_<name\*(C'\fR> will be set to the number of times that option was used.
.PP
If a flag option is specified multiple times, the value will be \f(CW\*(C`true\*(C'\fR. If a dual option is specified multiple times, the first value will be used. The \f(CW\*(C`option_count_<name\*(C'\fR> variable will still be accurate though.
.PP
Here's an example:
.PP
.Vb 5
\&    getopt_default=\-\-help \e
\&      getopt "$@" <<<"$spec_string"
\&    echo "The \*(Aq\-\-foo=\*(Aq option is \*(Aq$option_foo\*(Aq"
\&    echo "The \*(Aq\-\-bar\*(Aq option was used $option_count_bar times.
\&    echo "The remaining arguments are \*(Aq${args[*]}\*(Aq"
.Ve
.PP
\&\s-1NOTE:\s0 Bash has a builtin \f(CW\*(C`getopt\*(C'\fR command that doesn't do nearly as much. If you need to use the builtin in combination with this library, just use this command: \f(CW\*(C`builtin getopt ...\*(C'\fR for the builtin version.
.SS "\s-1SPEC STRUCTURE\s0"
.IX Subsection "SPEC STRUCTURE"
The \f(CW\*(C`getopt\*(C'\fR spec is a multiline string with 4 distinct sections:
.IP "1. Configuration Settings" 4
.IX Item "1. Configuration Settings"
You can put any number of lines at the top of the spec that have the form:
.Sp
.Vb 1
\&    getopt_config_var=value
.Ve
.Sp
These variables can be set elsewhere or not at all. Sometimes it is nice to put them in the spec since the spec acts as an overview of how the program works.
.IP "1. Usage Prose" 4
.IX Item "1. Usage Prose"
The first nonblank line(s) (after any config lines) is used as the \f(CW\*(C`usage\*(C'\fR in help or error messages. You should have one or more usage lines followed by a blank line.
.IP "1. Documentation and Commands" 4
.IX Item "1. Documentation and Commands"
The next section is where you put any free form documentation about how the command works. This is also where you typically put commands/descriptions if your program supports sub-commands. Like the \f(CW\*(C`git\*(C'\fR program has commands 'git clone\f(CW\*(C`and\*(C'\fRgit checkout\f(CW\*(C`, etc. This section ends with a mandatory \*(C'\fR\-\-` line.
.IP "1. Options Specification" 4
.IX Item "1. Options Specification"
Everything after the \f(CW\*(C`\-\-\*(C'\fR line is where the option parsing is defined in detail. It has its own very special format described in the next section.
.PP
The overall spec layout looks like:
.PP
.Vb 2
\&    getopt_default=\-\-help
\&    getopt_prefix=opt_
\&
\&    my\-program <opts> <args>
\&
\&    Description of my\-program.
\&
\&    Options:
\&    \-\-
\&    f,file=         File to use
\&    v,verbose       More output
.Ve
.SS "\s-1OPTIONS SPECIFICATION\s0"
.IX Subsection "OPTIONS SPECIFICATION"
To get good at using getopt-bash you should run \f(CW\*(C`git help rev\-parse\*(C'\fR, search for \f(CW\*(C`^PARSEOPT\*(C'\fR and read that entire section. The getopt-bash library uses \f(CW\*(C`git rev\-parse \-\-parseopt\*(C'\fR under the hood, but does a bunch of extra pre-processing and post-processing to make things really nice and simple.
.PP
The options section of the spec has these various combinations:
.PP
.Vb 5
\&    The options with \`str\` and \`num\` are enforced by getopt\-bash
\&    \-\-
\&    f,file=         Description of a value option (both short and long)
\&    help            Description of a flag option (long only)
\&    Q?              Description of a dual option (short only)
\&
\&    color=str       Value must be a string
\&    verbose?num     Value (if present) must be a number
.Ve
.PP
The getopt-bash library adds additional features to the spec options syntax:
.PP
.Vb 2
\&    input=+         This option is required
\&    size=+num       Required and must be a number
.Ve
.PP
In \f(CW\*(C`git rev\-parse \-\-parseopt\*(C'\fR there are 4 flag characters: \f(CW\*(C`= ? * !\*(C'\fR. The getopt-bash library adds:
.IP "\(bu" 4
\&\f(CW\*(C`+\*(C'\fR
.Sp
Option is required. This only makes sense for value options.
.PP
In \f(CW\*(C`git rev\-parse\*(C'\fR you can specify a hint word to indicate in the help text what type of value (\f(CW\*(C`num\*(C'\fR, \f(CW\*(C`str\*(C'\fR, etc) is expected for a value option.
.PP
Some hints are enforced and some are suggestions.
.PP
By default, getopt-bash will enforce the following hints:
.IP "\(bu" 4
\&\f(CW\*(C`str\*(C'\fR
.Sp
Value should be a string.
.IP "\(bu" 4
\&\f(CW\*(C`num\*(C'\fR
.Sp
Value should be a number.
.IP "\(bu" 4
\&\f(CW1..10\fR
.Sp
Value should be a number in the range 1\-10.
.IP "\(bu" 4
\&\f(CW\*(C`file\*(C'\fR
.Sp
Value should be a file that exists.
.IP "\(bu" 4
\&\f(CW\*(C`dir\*(C'\fR
.Sp
Value should be a directory the exists.
.IP "\(bu" 4
\&\f(CW\*(C`path\*(C'\fR
.Sp
Value should be the path of a file or directory that exists.
.IP "\(bu" 4
\&\f(CW\*(C`yyyy\-mm\-dd_HH:MM:SS\*(C'\fR
.Sp
Dates and times can be specified in any layout containing the strings: \f(CW\*(C`yyyy\*(C'\fR, \f(CW\*(C`yy\*(C'\fR, \f(CW\*(C`mm\*(C'\fR, \f(CW\*(C`dd\*(C'\fR, \f(CW\*(C`HH\*(C'\fR, \f(CW\*(C`MM\*(C'\fR, \f(CW\*(C`SS\*(C'\fR. Underscores are used for space characters. All other characters represent themselves.
.IP "\(bu" 4
\&\f(CW\*(C`/.../\*(C'\fR
.Sp
A hint enclosed in slash characters is a Bash regex to match.
.SS "\s-1CONFIGURATION\s0"
.IX Subsection "CONFIGURATION"
The \f(CW\*(C`getopt\*(C'\fR function has sensible defaults but it is also highly configurable. There are a number of \f(CW\*(C`getopt_...\*(C'\fR variables you can use. These variables can be set directly or by adding lines like:
.PP
.Vb 2
\&    getopt_default=(\-\-help)
\&    getopt_args=my_args
.Ve
.PP
to the beginning of your spec text.
.PP
The currently available option variables are:
.IP "\(bu" 4
\&\f(CW\*(C`getopt_default=...\*(C'\fR (default none)
.Sp
If no arguments were used, then use these. Multiple default arguments can be specified as an array: \f(CW\*(C`getopt_default=(\-\-foo \-\-bar)\*(C'\fR. This option is most commonly set to \f(CW\*(C`getopt_default=\-\-help\*(C'\fR (when appropriate).
.IP "\(bu" 4
\&\f(CW\*(C`getopt_debug=<debug\-opt\*(C'\fR> (default is \f(CW\*(C`debug\*(C'\fR)
.Sp
Defines which option turns on \f(CW\*(C`set \-x\*(C'\fR Bash debugging. The default is \f(CW\*(C`\-\-debug\*(C'\fR but that only works if you've defined the \f(CW\*(C`debug\*(C'\fR option in your spec. If you don't want to support this feature then just don't define \f(CW\*(C`debug\*(C'\fR or set this setting to an empty string.
.Sp
Note: This kind of debugging is extremely useful in tracking down hard to find problems.
.IP "\(bu" 4
\&\f(CW\*(C`getopt_prefix=word_\*(C'\fR (default is \f(CW\*(C`option_\*(C'\fR)
.Sp
The prefix for the getopt option variables where values are stored.
.IP "\(bu" 4
\&\f(CW\*(C`getopt_args=<var\*(C'\fR> (default is \f(CW\*(C`args\*(C'\fR)
.Sp
Name of the array variable to store the arguments found.
.IP "\(bu" 4
\&\f(CW\*(C`getopt_cmds=(...)\*(C'\fR (default is none)
.Sp
A list (array) of (sub\-)commands that have their own options.
.Sp
See \*(L"Commands and Options\*(R" below.
.IP "\(bu" 4
\&\f(CW\*(C`getopt_cmds_find=...\*(C'\fR (default is none)
.Sp
This config setting lets you define all commands that go into the \f(CW\*(C`getopt_cmds\*(C'\fR list in the options spec itself.
.Sp
If set to \f(CW\*(C`true\*(C'\fR it will take the first word from each line in the prose section that begins with spaces.
.Sp
Otherwise it is considered a regular expression to be applied to each line in the prose section. When a line matches, the first capture that is a word is added to the \f(CW\*(C`getopt_cmds\*(C'\fR list.
.IP "\(bu" 4
\&\f(CW\*(C`getopt_cmds_spec=<func\-name\*(C'\fR> (default is none)
.Sp
A callback function that writes an options spec for the \f(CW$cmd\fR command to stdout.

Changes
[version "0.1.24"]
date = 2023-01-01T13:03:27Z
done = Update .bpan/config
done = Version 0.1.23

[version "0.1.23"]
date = 2022-12-15T19:31:10Z
done = Add type to .bpan/config
done = Change package.name to package.id

[version "0.1.22"]
date = 2022-12-13-12-22-35
done = Config changes

[version "0.1.21"]
date = Mon Dec  5 12:49:20 PM PST 2022
done = Fix config

[version "0.1.20"]
date = Mon Dec  5 12:30:09 PM PST 2022
done = Change package.repo to package.source
done = Move Meta to .bpan/config

[version "0.1.19"]
date = Thu Nov 17 10:17:21 PM PST 2022
done = Update config to new format

[version "0.1.18"]
date = Wed Oct 26 06:47:30 AM PDT 2022
done = Revert "Make tests pass in Bash versions 3.2 - 5.2"

[version "0.1.17"]
date = Tue Oct 25 05:17:52 PM PDT 2022
done = Revert "Version 0.1.17"
done = Version 0.1.17
done = Make tests pass in Bash versions 3.2 - 5.2

[version "0.1.16"]
date = Tue Oct 18 10:28:41 AM PDT 2022
done = Refactor pager code
done = Change ReadMe.md from symlink to file

[version "0.1.15"]
date = Mon Oct 17 09:19:20 PM PDT 2022
done = Export the getopt function

[version "0.1.14"]
date = Wed Oct  5 08:02:56 PM PDT 2022
done = Fix .bpan/config symlink

[version "0.1.13"]
date = Wed Oct  5 08:16:19 AM PDT 2022
done = Support comment lines anywhere in spec

[version "0.1.12"]
date = Fri Sep 30 01:19:06 PM PDT 2022
done = Change 'update.template' to 'update.file' in config

[version "0.1.11"]
date = Fri Sep 30 08:17:03 AM PDT 2022
done = Support commenting out spec lines with '# '
done = Support '*' indicator for hidden option

[version "0.1.10"]
date = Wed Sep 28 09:12:18 AM PDT 2022
done = Make sure value options default being bound

[version "0.1.9"]
date = Sat Sep 24 11:20:53 AM EDT 2022
done = Update BPAN files
done = Support long name with dashes --foo-bar-baz

[version "0.1.8"]
date = Fri Sep 23 01:59:20 PM EDT 2022
done = Update to new config file schema

[version "0.1.7"]
date = Thu Sep 22 09:05:06 AM EDT 2022
done = Dual options default to '', not false
done = Update BPAN assets

[version "0.1.6"]
date = Thu Sep 15 03:07:46 PM EDT 2022
done = Remove all external deps (is-func)

[version "0.1.5"]
date = Tue Sep 13 04:33:03 PM EDT 2022
done = Fix bug in help output when parsing bad options

[version "0.1.4"]
date = Sun Sep 11 01:27:32 PM EDT 2022
done = Change test-more to test-tap

[version "0.1.3"]
date = Thu Sep  8 01:07:59 PM EDT 2022
done = Implement and document remaining functionality

[version "0.1.2"]
date = Mon Sep  5 07:56:27 AM PDT 2022
done = Finish implementing basic functionality and tests

[version "0.1.1"]
date = Tue 30 Aug 2022 07:30:57 AM PDT
done = Preserve blank lines in options usage output

[version "0.1.0"]
date = Tue 16 Aug 2022 09:22:29 AM PDT
done = Initial release


lib/getopt.bash
# getopt-long library based around `git rev-parse --parseopt`

getopt() {
  [[ ${getopt_spec-} ]] ||
    getopt_spec=$(cat)

  [[ $getopt_spec ]] ||
    getopt:error "No getopt 'spec' provided"

  getopt:set-vars

  getopt:parse-spec "$getopt_spec"

  [[ $# -gt 0 ]] ||
    set -- "${getopt_default[@]-}"

  if [[ ${#getopt_cmds[*]} -gt 0 ]]; then
    local words=()
    while [[ $# -gt 0 ]]; do
      if [[ " ${getopt_cmds[*]} " == *" $1 "* ]]; then
        cmd=$1
        shift
        words+=( -- "$@" )
        break
      else
        words+=( "$1" )
        shift
      fi
    done
    set -- "${words[@]}"
  fi

  local output
  local rc=0
  output=$(
    git rev-parse --parseopt --stuck-long -- "$@" <<<"$getopt_parseopt"
  ) || rc=$?

  if [[ $rc -ne 0 ]]; then
    if [[ $output == cat* ]]; then
      local pager=${PAGER-}
      [[ $pager && $pager != less ]] || pager='less -FRX'
      eval "$output" | $pager
    elif [[ $output ]]; then
      die "Unexpected results from 'git rev-parse --parseopt'"
    fi
    exit $rc
  else
    eval "$output"
  fi

  getopt:set-opts "$@"

  if [[ $cmd && $getopt_cmds_spec ]]; then
    getopt:cmd-getopt "${args[@]}"
  else
    local debug_var=$getopt_prefix$getopt_debug
    if [[ ${!debug_var-} == true ]]; then
      set -x
    fi
  fi
}

getopt:cmd-getopt() {
  [[ $(type -t "$getopt_cmds_spec") == function ]] ||
    getopt:error \
      "getopt_cmds_spec='$getopt_cmds_spec' must be a function name"
  local getopt_spec=''
  local getopt_default=()
  local getopt_cmds=()
  local getopt_cmds_find=''
  getopt_spec=$("$getopt_cmds_spec" "$cmd")

  if [[ $getopt_spec ]]; then
    local getopt_cmds_spec=''
    local cmd=''
    getopt "$@"
  fi
}

getopt:set-vars() {
  getopt_parseopt=''
  getopt_required=()
  getopt_database=''

  [[ ${getopt_cmds[*]+x} ]] || getopt_cmds=()
  [[ ${getopt_cmds_find+x} ]] || getopt_cmds_find=''
  [[ ${getopt_cmds_spec+x} ]] || getopt_cmds_spec=''

  cmd=''
  args=()
}

getopt:reset() {
  unset getopt_spec
  getopt_default=()
  getopt_debug=debug
  getopt_prefix=option_
  getopt_args=args

  cmd=''
  getopt_cmds=()
  getopt_cmds_find=''
  getopt_cmds_spec=''
}

getopt:source() {
  if [[ $# -eq 0 ]]; then
    getopt:error "Use '--' (source getopt.bash --) if no args for source."
  fi
  getopt:reset
  for arg; do
    if [[ $arg == -- ]]; then
      break
    elif [[ $arg == *$'\n--\n'* ]]; then
      getopt_spec=$arg
    else
      getopt:error "Invalid source arg '$arg'. Use '--' if no args."
    fi
  done
}

# Read through the provided 'getopt_spec' to collect the various instructions.
# Save all the instructions in a structured format in 'getopt_database'.
# Create a new spec 'getopt_parseopt' that is compliant with rev-parse's needs.
getopt:parse-spec() {
  local option_section=false
  local top=true
  local var count

  while IFS=$'\n' read -r line; do
    if [[ $line == \#\ * || $line == \# ]]; then
      continue
    fi

    # Look for getopt config variables at top of spec:
    if $top; then
      if [[ $line == getopt_* ]]; then
        eval "$line"
        continue
      elif [[ $line == '' ]]; then
        top=false
        continue
      else
        top=false
      fi
    fi

    # Check for start of options section:
    if ! $option_section; then
      if [[ $line == -- ]]; then
        option_section=true
      fi
      if [[ $getopt_cmds_find == true && $line == \ * ]]; then
        local cmd rest
        read -r cmd rest <<<"$line"
        getopt_cmds+=("$cmd")
      elif [[ $getopt_cmds_find && $line =~ $getopt_cmds_find ]]; then
        local match
        for match in "${BASH_REMATCH[@]}"; do
          if [[ $match =~ ^[-a-z0-9]+$ ]]; then
            getopt_cmds+=("$match")
          fi
        done
      fi
      getopt_parseopt+=$line$'\n'
      continue
    fi

    # Change an empty line to ' ' so it will appear in error messages:
    if [[ $line == '' ]]; then
      getopt_parseopt+=$' \n'
      continue
    fi

    # if 's,long...' (short,long)
    if [[ $line =~ ^([a-zA-Z0-9]),([a-z][-a-z0-9]+)([=?*][^\ ]*)?\  ]]; then
      short=${BASH_REMATCH[1]}
      long=${BASH_REMATCH[2]}
      kind=${BASH_REMATCH[3]:0:1}
      flag=${BASH_REMATCH[3]:1}
      name=$long

    # if 's...' (short only)
    elif [[ $line =~ ^([a-zA-Z0-9])([=?*][^\ ]*)?\  ]]; then
      short=${BASH_REMATCH[1]}
      long=''
      kind=${BASH_REMATCH[2]}
      flag=${BASH_REMATCH[2]:1}
      name=$short

    # if 'long...' (long only)
    elif [[ $line =~ ^([a-z][-a-z0-9]+)([=?*][^\ ]*)?\  ]]; then
      short=''
      long=${BASH_REMATCH[1]}
      kind=${BASH_REMATCH[2]}
      flag=${BASH_REMATCH[2]:1}
      name=$long

    else
      getopt:error "Bad line in getopt spec '$line'"
    fi

    var=$getopt_prefix${long:-$short}
    count=${getopt_prefix}count_${long:-$short}
    var=${var//-/_}
    count=${count//-/_}

    printf -v "$count" 0

    if [[ $kind == '='* ]]; then
      kind=value
      if ! [[ ${!var-} ]]; then
        # XXX Is there a better way to indirectly assign array variable =('')?
        if shopt -s compat41 2>/dev/null; then
          declare -g -a "$var"=''
        else
          eval "$var=('')"
        fi
      fi
    elif [[ $kind == '?'* ]]; then
      kind=dual
      printf -v "$var" ''
    else
      kind=bool
      if [[ ${!var-} != true ]]; then
        printf -v "$var" false
      fi
    fi

    type='' mult=false

    orig=flag
    while [[ $flag ]]; do
      if [[ ${flag:0:1} == '+' ]]; then
        getopt_required+=("$name")
        flag=${flag:1}
        line=${line/+/}
      elif [[ ${flag:0:1} == '*' ]]; then
        flag=${flag:1}
      elif [[ ${flag:0:1} == '@' ]] && ! $mult; then
        mult=true
        flag=${flag:1}
        line=${line/@/}
      elif [[ $flag =~ ^(str|num|file|dir|path) ]] ||
        [[ $flag =~ ^([0-9]+\.\.([0-9]+)?) ]]
      then
        type=${BASH_REMATCH[1]}
        flag=${flag:${#type}}
        line=${line/$type\ /\ }
      else
        getopt:error "Can't parse getopt flags '$orig' at '$flag'"
        flag=''
        exit 1
      fi
    done

    getopt_parseopt+=$line$'\n'

    getopt_database+="$name ${short:-_} ${long:=_} $kind ${type:-_} $mult"$'\n'
  done <<<"$1"
}

getopt:set-opts() {
  local option name var
  local need=" ${getopt_required[*]-}"
  local _name short long kind type mult

  option-vars() {
    name=${1#-}
    name=${name#-}
    unset value
    if [[ $name == *=* ]]; then
      value=${name#*=}
      name=${name%%=*}
    fi
    spec=$(grep "^$name " <<<"$getopt_database") ||
      die "Can't grep '^$name ' in:" "$getopt_database"
    read -r _name short long kind type mult <<<"$spec"
    long=${long#_}
    short=${short#_}
    show=${long:+"--$long=…"}
    show=${show:-"-$short …"}
  }

  while [[ $# -gt 0 ]]; do
    option=$1; shift

    if [[ $option == -- ]]; then
      if [[ ${need##\ } ]]; then
        for option in $need; do
          option-vars "$option"
          echo "* Option '$show' is required" >&2
        done
        getopt:error 'Missing required options'
      fi

      # TODO use getopt_args here
      args=("$@")
      return
    fi

    option-vars "$option"

    need=${need/\ $name/}

    var=$getopt_prefix${long:-$short}
    count=${getopt_prefix}count_${long:-$short}
    var=${var//-/_}
    count=${count//-/_}

    # Set bool option variable to 'true':
    if [[ $kind == bool ]]; then
      printf -v "$var" true

    # Push value onto option array:
    elif [[ $kind == value ]]; then
      [[ $type ]] && getopt:validate
      IFS=$'\n' read -r -d '' "${var}[${!count}]" <<<"$value" || true

    elif [[ $kind == dual ]]; then
      if [[ ${value-} ]]; then
        [[ $type ]] && getopt:validate
        if ! [[ ${!var-} ]]; then
          printf -v "$var" '%s' "$value"
        fi
      else
        printf -v "$var" true
      fi
    else
      die
    fi

    # Increment option_count_<name>
    # shellcheck disable=2004
    : $(( ${count} = ${!count} + 1 ))
  done

  [[ $# -eq 0 ]] ||
    getopt:error "Unexpected arguments: '$*'"
}

getopt:validate() (
  case "$type" in
    _) ;;
    str) ;;
    file) [[ -f "$value" ]] ||
      getopt:error "$(getopt:msg) file does not exist";;
    dir) [[ -d "$value" ]] ||
      getopt:error "$(getopt:msg) directory does not exist";;
    path) [[ -e "$value" ]] ||
      getopt:error "$(getopt:msg) path does not exist";;
    *)
      if [[ $type == num ]] &&
         [[ ! $value =~ ^-?[0-9]+$ ]]
      then
        getopt:error "$(getopt:msg) is not a number"
      elif [[ $type =~ ^([0-9]+)\.\.([0-9]+)$ ]]; then
        lower=${BASH_REMATCH[1]}
        upper=${BASH_REMATCH[2]}
        if [[ ! $value =~ ^-?[0-9]+$ ]] ||
           [[ $value -lt $lower ]] ||
           [[ $value -gt $upper ]]
        then
          getopt:error "$(getopt:msg) is not in the range"
        fi
      else
        getopt:error "Validation type '$type' not yet supported"
      fi
      ;;
  esac
)

getopt:msg() (
  echo "$option is type '$type', but value '$value'"
)

getopt:error() {
  die "[getopt] Error: $1"
}

[[ $0 == "${BASH_SOURCE[0]}" ]] ||
  getopt:source "$@"

export -f getopt
